outlines/base.py

```diff
@@ -0,0 +1,163 @@
+import asyncio
+import inspect
+
+import numpy as np
+
+
+class vectorize:
+    """Returns an object that acts like a function but takes arrays as an input.
+
+    The vectorized function evaluates `func` over successive tuples of the input
+    chararrays and returns a single NumPy chararrays or a tuple of NumPy chararrays.
+
+    Its behavior is similar to NumPy's `vectorize` for Python functions: the function
+    being vectorized is executed in a `for` loop. Coroutines, however, are executed
+    concurrently.
+
+    Part of the code was adapted from `numpy.lib.function_base`.
+
+    """
+
+    def __init__(self, func, signature=None):
+        self.func = func
+        self.signature = signature
+        self.is_coroutine_fn = inspect.iscoroutinefunction(func)
+
+        self.__name__ = func.__name__
+        if hasattr(func, "__doc__"):
+            self.__doc__ = func.__doc__
```

Comment:
This makes me wonder if some `functools.wraps`-based approach would work for this as well.

Reply:
Replaced these lines with `functools.update_wrapper(self, func)`.